src/parquet/arrow/arrow-reader-writer-test.cc

```diff
@@ -993,6 +995,145 @@ TEST(TestArrowReadWrite, DateTimeTypes) {
+      s2, ColumnVector({std::make_shared<Column>("f_s", a_ms),
+                        std::make_shared<Column>("f_ms", a_ms),
+                        std::make_shared<Column>("f_us", a_ms),
+                        std::make_shared<Column>("f_ns", a_ms)}));
+
+  // Result when coercing to microseconds
+  auto s3 = std::shared_ptr<::arrow::Schema>(
+      new ::arrow::Schema({field("f_s", t_us), field("f_ms", t_us), field("f_us", t_us),
+                           field("f_ns", t_us)}));
+  auto ex_micro_result = std::make_shared<::arrow::Table>(
+      s3, ColumnVector({std::make_shared<Column>("f_s", a_us),
+                        std::make_shared<Column>("f_ms", a_us),
+                        std::make_shared<Column>("f_us", a_us),
+                        std::make_shared<Column>("f_ns", a_us)}));
+
+  std::shared_ptr<Table> milli_result;
+  DoSimpleRoundtrip(
+      input, 1, input->num_rows(), {}, &milli_result,
+      ArrowWriterProperties::Builder().coerce_timestamps(TimeUnit::MILLI)->build());
+  ASSERT_TRUE(milli_result->Equals(*ex_milli_result));
+
+  std::shared_ptr<Table> micro_result;
+  DoSimpleRoundtrip(
+      input, 1, input->num_rows(), {}, &micro_result,
+      ArrowWriterProperties::Builder().coerce_timestamps(TimeUnit::MICRO)->build());
+  ASSERT_TRUE(micro_result->Equals(*ex_micro_result));
+}
+
+TEST(TestArrowReadWrite, CoerceTimestampsLosePrecision) {
+  using ::arrow::ArrayFromVector;
+  using ::arrow::field;
+
+  // PARQUET-1078, coerce Arrow timestamps to either TIMESTAMP_MILLIS or TIMESTAMP_MICROS
+  std::vector<bool> is_valid = {true, true, true, false, true, true};
+
+  auto t_s = ::arrow::timestamp(TimeUnit::SECOND);
+  auto t_ms = ::arrow::timestamp(TimeUnit::MILLI);
+  auto t_us = ::arrow::timestamp(TimeUnit::MICRO);
+  auto t_ns = ::arrow::timestamp(TimeUnit::NANO);
+
+  std::vector<int64_t> s_values = {1489269, 1489270, 1489271, 1489272, 1489272, 1489273};
+  std::vector<int64_t> ms_values = {1489269001, 1489270001, 1489271001,
+                                    1489272001, 1489272001, 1489273001};
+  std::vector<int64_t> us_values = {1489269000001, 1489270000001, 1489271000001,
+                                    1489272000001, 1489272000001, 1489273000001};
+  std::vector<int64_t> ns_values = {1489269000000001LL, 1489270000000001LL,
+                                    1489271000000001LL, 1489272000000001LL,
+                                    1489272000000001LL, 1489273000000001LL};
+
+  std::shared_ptr<Array> a_s, a_ms, a_us, a_ns;
+  ArrayFromVector<::arrow::TimestampType, int64_t>(t_s, is_valid, s_values, &a_s);
+  ArrayFromVector<::arrow::TimestampType, int64_t>(t_ms, is_valid, ms_values, &a_ms);
+  ArrayFromVector<::arrow::TimestampType, int64_t>(t_us, is_valid, us_values, &a_us);
+  ArrayFromVector<::arrow::TimestampType, int64_t>(t_ns, is_valid, ns_values, &a_ns);
+
+  auto s1 = std::shared_ptr<::arrow::Schema>(new ::arrow::Schema({field("f_s", t_s)}));
```

Comment:
I created ARROW-1336; I think this could be solved with a public factory function